Let SFTP server figure out remote users home directories

capnspacehook opened the pull request:
Previously a Teleport client using SFTP would resolve remote host user
home directories by making a subsystem request to a Teleport server
which would return the home directory. The problem was the subsystem
request counted as an open session, which could make the SFTP file
transfer fail. This was frustrating and didn't make much sense, but
after reading the SFTP specification again I realized that SFTP servers
are to handle relative paths by assuming they start at the user's home
directory. So let the server figure out the correct path and remove any
tilde prefixes from remote paths.

Fixes https://github.com/gravitational/teleport/issues/23593.

russjones reviewed (commented):
Why a new channel for this?

The SSH protocol supports sending requests within a channel and over a global channel. We use that approach in other places where we need to get a string back (like version information).

https://github.com/gravitational/teleport/blob/master/lib/reversetunnel/srv.go#L1242
https://github.com/gravitational/teleport/blob/master/lib/reversetunnel/agent.go#L461

That feels like a better fit than creating a whole new channel for a string. What am I missing?

espadolini commented:
What does opensshd sftp do when connecting to openssh sshd, to resolve `~`?

+1 for using a global request rather than opening a channel. It might make even more sense as a request local to the sftp subsystem channel, but sadly those can't carry a payload back so you'd need to either always blindly send it from the server side (and just wait with a timeout on the client side) or do some awkward back and forth with channel requests that are technically unrelated, and for a long time we've just been killing the channel whenever the server side received an unknown channel request (yay for extensibility, huh) so we can't really extend that part of the protocol until that's fixed and we pass a couple of major versions.

capnspacehook committed:
let SFTP server figure out remote users home directories

Previously a Teleport client using SFTP would resolve remote host user
home directories by making a subsystem request to a Teleport server
which would return the home directory. The problem was the subsystem
request counted as an open session, which could make the SFTP file
transfer fail. This was frustrating and didn't make much sense, but
after reading the SFTP specification again I realized that SFTP servers
are to handle relative paths by assuming they start at the user's home
directory. So let the server figure out the correct path and remove any
tilde prefixes from remote paths.

russjones commented on lib/sshutils/sftp/sftp.go:
```diff
@@ -199,94 +196,85 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
+		return 1
+	}
+
+	return -1
```
Wouldn't returning 0 make more sense here? Then in `expandPath` you can just do:

```go
n := homeDirPrefixLen(pathStr)
return pathStr[n:], nil
```

capnspacehook replied:
Not really, as I need to check for other edge cases anyway, so this wouldn't buy me much.

capnspacehook replied:
Good point, done

russjones commented on lib/sshutils/sftp/sftp.go:
```diff
@@ -199,94 +196,85 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
+		return 2
 	}
-	return strings.HasPrefix(path, "~/")
+	if path[0] == '~' {
+		return 1
+	}
```
@capnspacehook What does OpenSSH do in this situation?

capnspacehook replied:
You're right, changed to only expand `~user` paths when they are local and return an error otherwise. Expanding remote paths in this way requires SFTP server support, I could easily implement it in a Teleport-specific way but I'm leaning towards not alienating OpenSSH `scp` users.

I might try and open an upstream PR that adds the ability for our SFTP server to do this, but it's complicated as OpenSSH `scp` will try and use a OpenSSH specific request to resolve home dirs, while most other SFTP clients will use a different standardized request. I may have to implement both, but idk if upstream maintainers will want that.

capnspacehook replied:
I'll tackle that in another PR though, and we already have an [issue to track it](https://github.com/gravitational/teleport/issues/22886) so I'm going to mark this resolved now.

russjones commented on lib/sshutils/sftp/sftp.go:
```diff
@@ -199,94 +196,85 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
+		}
+		// this is safe because we verified that all paths are non-empty
+		// in CreateUploadConfig/CreateDownloadConfig
+		return path.Join(homeDir, pathStr[1:]), nil
```
You're not using `pfxLen` here, is that intentional?

capnspacehook replied:
It was, but I'm using `pfxLen` here now due to the logic change.

russjones commented on lib/sshutils/sftp/sftp.go:
```diff
@@ -323,94 +320,112 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
+		// if the path starts with ~user, resolve that user's home dir
+		// otherwise resolve the current user's home dir
+		if pfxLen == 1 && len(pathStr) > 1 {
+			name, _, ok := strings.Cut(pathStr, "/")
+			// if we didn't find anything with back slashes, try with
+			// forward slashes when on Windows
+			if runtime.GOOS == "windows" && !ok {
+				name, _, ok = strings.Cut(pathStr, `\`)
+			}
```
Can you use `os.PathSeparator` here instead, it will consolidate these lines.

capnspacehook replied:
I'm doing this on purpose, as Windows will (mostly) allow `\` or `/` as path separators, so I allow the user to specify a path either way as most Windows tools and the OS itself does.

capnspacehook replied:
Moot now though, removed this code

russjones commented on lib/sshutils/sftp/sftp.go:
```diff
@@ -323,94 +320,112 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
+		}
+		// this is safe because we verified that all paths are non-empty
+		// in CreateUploadConfig/CreateDownloadConfig
+		return path.Join(homeDir, pathStr[pfxLen:]), nil
```
I'd use `filepath.Join` instead, it will be another situation where you don't have to think about OS specific separators.

capnspacehook replied:
SFTP requires paths separated by `/`, so the thought was to just use `path.Join` here so that's enforced. It doesn't really matter since paths are `path.Clean`ed anyway. Regardless, this code is now removed given your comment below.

russjones commented on lib/sshutils/sftp/sftp.go:
```diff
@@ -323,94 +320,112 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
-		return pathStr, nil
+	// if the path is local, expand the path to be rooted at this user's
+	// home dir
+	if !isRemotePath {
```
A general question about this block. Typically the shell will expand `~`. Are we expanding `~` here for local paths incase a shell was not used to invoke `tsh scp`?

I think our guiding light here should be OpenSSH interoperability. Will OpenSSH ever expand `~` for local paths?

capnspacehook replied:
That's a good point, OpenSSH `scp` does not expand local paths so I'll remove that code

espadolini commented on lib/sshutils/sftp/sftp.go:
```diff
@@ -322,95 +313,70 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
-	homeDir, err := getHomeDir()
-	if err != nil {
-		return "", trace.Wrap(err)
+	// Removing the home dir prefix would mean returning an empty string,
+	// which is supported by SFTP but won't be as clear in logs or audit
+	// events. Since the SFTP server will be rooted at the user's home
+	// directory, "." and "" are equivalent in this context.
```
Should we handle `""` in the same way as `"~"` then? Can we ever get an empty `pathStr` argument here?

capnspacehook replied:
We can't get an empty string here, `CreateUploadConfig` and `CreateDownloadConfig` ensure that the source and destination paths are not empty, and the `Config` struct's important fields are not exported meaning users can't construct it themselves

capnspacehook commented:
Friendly ping @xacrimon

r0mant commented:
@capnspacehook Can you backport this to branch/v12 as soon as you can please?
